submissions/UA-girl/task_friends_app/js/app.js

```diff
@@ -0,0 +1,300 @@
+        this.sortingStates = {
+            name: null,
+            registration: null,
+            age: null
+        };
+    }
+
+    genderFilter(gender, wasToggled = true) {
+        this.filterStates.gender = gender;
+        if (wasToggled) {
+            this.restoreFromFilter();
+        } else {
+            this.cards = this.cards.filter(person => person.gender === gender);
+        }
+    }
+
+    ageFilter(maxAge) {
+        document.querySelector('.label-range').innerHTML = `Max age: ${maxAge}`;
+        if (this.filterStates.age < +maxAge) {
+            this.filterStates.age = +maxAge;
+            this.restoreFromFilter();
+        } else {
+            this.filterStates.age = +maxAge;
+            this.cards = this.cards.filter(person => person.age <= +maxAge);
+        }
+    }
+
+    sortCards(value, filterName) {
+        this.resetSortingState();
+        switch (filterName) {
+            case 'age':
+                this.cards.sort((person1, person2) => person1.compareAge(person2));
+                break;
+            case 'name':
+                this.cards.sort((person1, person2) => person1.compareName(person2));
+                break;
+            case 'registration':
+                this.cards.sort((person1, person2) => person1.compareDateRegistration(person2));
+        }
+        this.sortingStates[filterName] = value;
+        if (value === 'down') {
+            this.cards.reverse()
+        }
+    }
+
+    nameFilter(name) {
+        name = name.trim().toLowerCase();
+        this.restoreFromFilter();
+        this.cards = this.cards.filter(person => person.name.first.toLowerCase()
+            .startsWith(name) || person.name.last.toLowerCase().startsWith(name));
+    }
+
+    disableEnablePageToggler() {
+        const [backPage, forwardPage] = document.querySelectorAll('.button-page');
+        if ((this.currentPage + 1) * CARDS_PER_PAGE >= this.cards.length) {
+            forwardPage.classList.add('disabled');
+            forwardPage.setAttribute('disabled', true);
+        } else {
+            forwardPage.removeAttribute('disabled');
+            forwardPage.classList.remove('disabled');
+        }
+        if (this.currentPage === 0) {
+            backPage.classList.add('disabled');
+            backPage.setAttribute('disabled', true);
+        } else {
+            backPage.classList.remove('disabled');
+            backPage.removeAttribute('disabled');
+        }
+        const span = document.querySelector('.page');
+        span.textContent = this.currentPage + 1;
+    }
+
+    resetApp() {
+        this.cards = cards.slice();
+        this.resetFilterState(true);
+        this.resetSortingState();
+        document.querySelector('.label-range').innerHTML = `Max age: ${MAX_AGE}`;
+        document.querySelector('.range').value = MAX_AGE;
+        document.querySelector('.search').value = '';
+        document.querySelectorAll('.radio').forEach(input => input.checked = false)
```

Comment:
You can move querying of these elements to the constructor in order to prevent calling DOM assertions multiple times

Reply:
Thank you! Do you mean, that I can create class variables, and here just update these elements? Because here I want to completely rub off the all filters and sorting, when user makes a reset.

Reply:
yes, just create these variables in the constructor:
``` javascript
this.rangeLabel = document.querySelector('.label-range');
this.range = document.querySelector('.range')
....
```
It would help you to store them in cache and not to use `document.querySelector` on each calling the `resetApp` method

Reply:
Well, I've got it, thank you!